### ISSUE TITLE
starting with server.port -1 will stop all connectors not only the http ones 

### ISSUE THREAD

**Issue by masrawi** [spring-projects/spring-boot#3359]
I have a situation where I want to disable all http connectors and enable only an AJP one. the AJP connector only starts if the server port is >= 0. looking at the code it seems that all connectors are stopped once the port is < 0

**Comment by philwebb** [comment]
Can you provide some more details here?

> looking at the code it seems that all connectors are stopped once the port is < 0

What code are you looking at?

You might need to use a `TomcatConnectorCustomizer` bean if you want fine grain control over individual connectors.

**Comment by masrawi** [comment]
TomcatEmbeddedServletContainerFactory has the method 
```
protected TomcatEmbeddedServletContainer getTomcatEmbeddedServletContainer(
			Tomcat tomcat) {
		return new TomcatEmbeddedServletContainer(tomcat, getPort() >= 0);
	}
``` 
which starts the container with autostart = false in case of port -1 

the start method
```
public void start() throws EmbeddedServletContainerException {
		addPreviouslyRemovedConnectors();
…
```  
will remove _ALL_ the connectors when started  _including_ the AJP one

```
private void addPreviouslyRemovedConnectors() {
		Service[] services = this.tomcat.getServer().findServices();
		for (Service service : services) {
			Connector[] connectors = this.serviceConnectors.get(service);
			if (connectors != null) {
				for (Connector connector : connectors) {
					service.addConnector(connector);
					if (!this.autoStart) {
						stopProtocolHandler(connector);
					}
				}
				this.serviceConnectors.remove(service);
			}
		}
	}
```

in my case I want to disable _only_ the http connectors as the docs states

**Comment by wilkinsona** [comment]
Why are you setting the port to -1? Why not use a `TomcatConnectorCustomizer`, as @philwebb suggested, to turn the default connector into an AJP one?

**Comment by masrawi** [comment]
if you mean like this.
```
@Bean
    @ConditionalOnExpression("${http.connector.off}")
    public EmbeddedServletContainerCustomizer containerCustomizer()
    {
        return new AjpCusotomizer(environment.getProperty("ajp.port"));
    }

    private static class AjpCusotomizer implements EmbeddedServletContainerCustomizer
    {

        private String ajpport;

        public AjpCusotomizer(String port)
        {
            ajpport = port;
        }

        @Override
        public void customize(ConfigurableEmbeddedServletContainer factory)
        {
            if (factory instanceof TomcatEmbeddedServletContainerFactory)
            {
                ((TomcatEmbeddedServletContainerFactory) factory).setProtocol("AJP/1.3");
                customizeTomcat((TomcatEmbeddedServletContainerFactory) factory);
            }
        }

        public void customizeTomcat(TomcatEmbeddedServletContainerFactory factory)
        {
            factory.addConnectorCustomizers(new TomcatConnectorCustomizer() {

                @Override
                public void customize(Connector connector)
                {
                    connector.setProtocolHandlerClassName(AjpNioProtocol.class.getName());
                    connector.setProtocol("AJP/1.3");
                    connector.setPort(Integer.parseInt(ajpport));
                    connector.setEnableLookups(false);
                   
                }
            });
        }
    }
```
it does not work and starts an http connector instead of the AJP one. the point is that the implementation will stop all the connectors instead of only the http ones

**Comment by wilkinsona** [comment]
As I said above, "why are you setting the port to -1?". If you want an AJP connector instead of an HTTP one, why not use `server.port` to configure its port and a `TomcatConnectorCustomizer` to change the connector's protocol from HTTP to AJP?

**Comment by wilkinsona** [comment]
Perhaps you're being mislead by the log entry like this:

```
2015-07-02 13:58:31.035  INFO 19039 --- [           main] s.b.c.e.t.TomcatEmbeddedServletContainer : Tomcat started on port(s): 8080 (http)
```

Note that the `http` is in reference to the connector's scheme, not its protocol.

**Comment by masrawi** [comment]
the above code I wrote is based on server.port 0 but for some reason I am still able to call the browser with http. it seems the connector type is not affected

**Comment by wilkinsona** [comment]
This works for me:

```java
@Bean
public EmbeddedServletContainerCustomizer servletContainerCustomizer() {
	return new EmbeddedServletContainerCustomizer() {

		@Override
		public void customize(ConfigurableEmbeddedServletContainer container) {
			if (container instanceof TomcatEmbeddedServletContainerFactory) {
				((TomcatEmbeddedServletContainerFactory) container)
						.setProtocol("AJP/1.3");
			}
		}

	};
}
```

curl  fails to connect due to an empty response:

```
$ curl localhost:8080
curl: (52) Empty reply from server
```

If you'd like me to spend some more time on this, please share a sample that reproduces the problem.

**Comment by masrawi** [comment]
Hi Andy, thanks your code works for me as well. So I guess the docs should be clearer about the -1 port and may be your code about overriding the default connector.

cheers

**Comment by shakuzen** [comment]
I know this issue is closed now, but I just thought I would share that I ran into the same problem as @masrawi and was confused in the same way by the documentation that states `server.port=-1` will turn off all HTTP endpoints. This is from [67.3 Change the HTTP port](http://docs.spring.io/spring-boot/docs/1.3.0.BUILD-SNAPSHOT/reference/htmlsingle/#howto-change-the-http-port), which I see is unchanged.

Depending on the environment, I wanted to make an HTTP port **and/or** an AJP port configurable via properties. My first attempt led me to add a property for the AJP port (e.g. `mycom.server.ajp-port`). Then the way I thought one would indicate not wanting an HTTP connector would be by specifying `server.port=-1` (since `server.port` is for the default HTTP connector) but as stated in this issue, that will disable any ports you configure. This was not my expectation.

Locally, where we use only an HTTP port for testing, everything worked fine. Then in our servers where we typically use AJP only or AJP and HTTP, things were not working (in the AJP only case).

If `server.port` wasn't set and `mycom.server.ajp-port` was set, I suppose I could have customized the existing connector, but then that sort of overwrites the existing default functionality of an HTTP port 8080 connector when `server.port` wasn't set, which I wanted to preserve for local testing.

Eventually, I determined we did not really need the case of both an AJP and HTTP connector configured at the same time and so I replaced the `mycom.server.ajp-port` property with one like `mycom.server.connector.protocol` that defaults to `HTTP/1.1`. This way, from properties an AJP connector could be configured without breaking or changing any of the default functionality.

But, it did cause significant confusion and investigation to realize `server.port=-1` was causing ALL connectors to be disabled. I just wanted to share my experience. I don't know how many other people are actually needing to use an AJP connector and running into this kind of issue.
